Small improvements to NJT matrix multiplies

michael-diggin opened the pull request:
Fixes #146404

Adds changes to the matmul and matmul_backward operation for nested jagged tensors, to support back propagation when the output is a regular strided tensor.
This required adding support for the nested matmul operation to work when the nested tensor wasn't 'self', i.e
`A@B` where `A` isn't nested but `B` is.

The operation schemas had to be updated to reflect that either input can be a strided tensor instead (and the gradient), so an extra assertion is added in an edge case where neither input is nested.

Unit tests are also added.

michael-diggin committed:
Small improvements to NJT matrix multiplies

michael-diggin commented:
@pytorchbot label "topic: not user facing"

michael-diggin committed:
fix device use in test

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -2559,6 +2587,14 @@ def matmul_backward_default(func, *args, **kwargs):
+    if not inp.is_nested and not other.is_nested:
+        raise ValueError(
+            "matmul_backward_default: expected one of [self, other] to be a nested tensor"
+        )
```
You can't actually get to this `matmul_backward_default` function  if neither of the inputs are nested tensors

michael-diggin replied:
Yep true, removed - thanks!

michael-diggin committed:
remove check from matmul_backward_default

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -2247,6 +2247,38 @@ def test_matmul_noncontiguous(self, device, dtype):
 
+    @dtypes(torch.float, torch.double)
+    def test_mamtul_dense_x_njt(self, device, dtype):
+        # T (B, D, E) with NJT (B, *, E, C) broadcasting case
```
random_nt_from_dims([3, None, 4]) produces (B, *, E) not (B, *, E, C) right

michael-diggin replied:
Yes, I had forgotten to update the comment, I've removed this test anyway now for the OpInfo one

soulitzer reviewed (commented):
Thanks for the PR! Added some small comments

Btw, there's some existing infra around testing that we can use that will automatically test forward, backward, compile, etc. (See the tests under TestNestedTensorOpInfo)

To test the new inputs that you are supporting here, instead of adding new one-off tests here, you can add the inputs that you would like to test to `sample_inputs_matmul` in `torch/testing/_internal/opinfo/definitions/nested.py`

michael-diggin committed:
small chanages and use test infra

michael-diggin commented on torch/nested/_internal/ops.py:
```diff
@@ -1153,14 +1153,23 @@ def matmul_default(func, *args, **kwargs):
+        raise ValueError(
+            "matmul_default: expected one of [self, other] to be a nested tensor"
+        )
+
```
Removed

michael-diggin commented on torch/nested/_internal/ops.py:
```diff
@@ -1202,6 +1215,21 @@ def _padded_impl(a, b):
+        elif inp.dim() == 2 and other.dim() > inp.dim():
+            return NestedTensor(
+                func(inp, other._values, **new_kwargs), **extract_kwargs(other)
+            )
```
Good call, added this to the various if clauses.

michael-diggin commented:
> Thanks for the PR! Added some small comments
> 
> Btw, there's some existing infra around testing that we can use that will automatically test forward, backward, compile, etc. (See the tests under TestNestedTensorOpInfo)
> 
> To test the new inputs that you are supporting here, instead of adding new one-off tests here, you can add the inputs that you would like to test to `sample_inputs_matmul` in `torch/testing/_internal/opinfo/definitions/nested.py`

Thanks for the review @soulitzer! I've made changes based on the comments, and updated the tests to use the existing test infra/setup which is much nicer.

soulitzer commented on torch/testing/_internal/opinfo/definitions/nested.py:
```diff
@@ -1021,6 +1021,53 @@ def sample_inputs_matmul(
+        # (B, E, j1) x (B, j1, F) => (B, E, F)
+        if njt_3d._ragged_idx == 2 and njt_3d.is_contiguous():
+            B, E, _ = njt_3d.shape
+            j1 = len(njt_3d.values())
```
```suggestion
            sum_j1 = len(njt_3d.values())
```
nit: Usually j1 refers to the ragged dimension, but what you are trying to express here is the entire length of the packed dimension, e.g. sum(j1)

soulitzer reviewed (approved):
Thanks for quick updates! Just have another nit

michael-diggin committed:
rename var to sum_j1

michael-diggin commented:
> Thanks for quick updates! Just have another nit

Thanks for the quick review @soulitzer! Just committed that small change. Could you rerun the workflows when you get a chance? As I think that's needed before I can merge.

jbschlosser reviewed (approved):
Awesome, thanks for the PR! looks solid on my end, thanks for updating the generated sample inputs to cover this case :)

soulitzer commented:
@pytorchbot merge
